docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+
+*Rationale:* This case is reached if we are packing a union type inside the significand bits of a signalling NaN and guarantees that we can load the payload in a single aligned load and need to zero bits at only one end to extract the payload.
+
+In other cases, the discriminator bits are stored in the low bits (as defined by the platform byte order) of the object and the payload in the high bits.
+
+*Rationale:* The discriminator is easy to mask off with a right shift and extract with an xor with an immediate operand for comparisons.
+
+The types identified in a discriminator are sorted lexically and discriminator values are assigned in order.
+Singletons are sorted before all other types.
+
+*Rationale:* The representation of `A | B` and `B | A` should be the same.
+It is useful to sort singletons before pointers so that the common case of `T* | NotAThing` is able to use null-pointer checks and branches, which are typically optimised aggressively in modern microarchitectures.
+
+{*Note:* I'm not sure lexical sorting will work because types imported from the same package by two other packages will have different names and so we may need some other stable ordering for types.}
+
+*Open question*: As described, this requires a pointer dereference do disambiguate pointer types.
+It would be possible to use any space in the low three bits that is not used to disambiguate non-pointer types to differentiate a small number of pointee types.
+This would save a load in pattern matching and might be useful, though given that the main reason for pattern matching is to then call a method on the concrete type, it's not clear that this is actually a win.
+Note that, in the CHERI case, the low bits are still available and so it may be useful to use them to differentiate between pointee types.
+
+These rules are somewhat complex in the abstract.
+The following sections give some concrete examples of how they are applied.
+
+### Example `U64 | Float64`
+
+The size of `U64` is 64 bits and it requires one discriminator state.
+The size of `Float64` is 64 bits and it requires one discriminator state.
+
+We first try the special-case rule for the case that the largest value is a floating-point value.
+The remaining type (`U64`) requires 64 bits for payload and 0 bits for discriminator, and so does not fit into the significand part of a `Float64` signalling NaN.
+The first special case is therefore not used
+The second special case is not used because neither type is a pointer.
+
+We are now left requiring one bit of discriminator and 64 bits of payload.
+The payload size is a power of two, and so the address of the union type is the address of the start of a 64-bit payload.
+The 1-bit discriminator is stored afterwards, with a value of 0 representing a `Float64` and a value of `1` representing a `U64`.
+The field-packing algorithm is free to use the remaining 7 bits in the discriminator for other values.
+
```

Comment:
Possibly worth adding, just for completeness, that, in this example, the alignment requirement is 8 bytes and so there are 7 bytes (and 7 bits) of padding.

Reply:
I don't want to talk about padding here, because padding is a thing that happens in aggregate types and there are more complex rules for that.